File: Go/Euler012.go
package main

import (
	"fmt"
	"math"
)

func main() {
	const target = 500

	triangle := 0
	incrementer := 1
	for len(getDivisors(triangle)) < target {
		triangle += incrementer
		incrementer++
	}

	fmt.Println(triangle)
}

func getDivisors(num int) map[int]struct{} {
	max := int(math.Ceil(math.Sqrt(float64(num))))
	divisors := make(map[int]struct{})

	for i := 2; i <= max; i++ {
		if num%i == 0 {
			divisors[i] = struct{}{}
			divisors[num/i] = struct{}{}
		}
	}

	return divisors
}

File: Kotlin/Euler012.kt
package Euler012

import java.lang.Math

fun main() {
  val target = 500

  var triangle = 0

  var incrementer = 1
  while (true) {
    triangle += incrementer++
    val factors = getFactors(triangle)
    if (factors.size >= target) {
      println(triangle)
      break
    }
  }
}

fun getFactors(n: Int): Set<Int> {
  var factors: MutableSet<Int> = mutableSetOf()

  val max = Math.sqrt(n.toDouble()).toInt()
  for (f in 1..max) {
    if (n % f != 0) continue

    factors.add(f)
    factors.add(n / f)
  }

  return factors
}
